drivers/serial/uart_nrfx_uart.c

```diff
@@ -414,6 +419,31 @@ static int uart_nrfx_callback_set(const struct device *dev,
 	return 0;
 }
 
+/* If there's a timeout on transmission completion, schedule it to be
+ * processed.
+ */
+static inline void tx_timeout_schedule(void)
+{
+#if HW_FLOW_CONTROL_AVAILABLE
+	if (uart0_cb.tx_timeout != SYS_FOREVER_MS) {
+		atomic_set_bit(uart0_cb.pending_flags, PENDING_TX_TIMEOUT_BIT);
+		k_work_schedule(&uart0_cb.tx_timeout_work,
+				K_MSEC(uart0_cb.tx_timeout));
+	}
+#endif
+}
+
+/* Attempt to cancel a pending timeout on transmission completion. */
+static inline void tx_timeout_cancel(void)
+{
+#if HW_FLOW_CONTROL_AVAILABLE
+	if (uart0_cb.tx_timeout != SYS_FOREVER_MS) {
+		k_work_cancel(&uart0_cb.tx_timeout_work);
+		atomic_clear_bit(uart0_cb.pending_flags, PENDING_TX_TIMEOUT_BIT);
```

Comment:
this will not work as intended if timeout is restarted quickly
e.g.
* schedule first timeout
* ... time pass ..
* work is submitted (i.e. cannot be canceled anymore)
* cancel timeout (clears atomic in hope callback will notice it)
* schedule new timeout (bit is set)
* callback (for the original timeout) is executed by a workqueue and sees bit set
* timeout is triggered incorrectly (too soon)

If work had a private/magic it could be tagged but since it does not have a private data tagging is impossible.
Possible option is to use be ref counting on pending submissions. Only when reaching zero actual timeout code should be executed.

Reply:
Unless timeouts are always long enough to make sure that such race will not happen.

Reply:
OK, I've dropping that commit and somebody else can figure out how to fix #29173 in another PR.  Please remove your NAK if this is your only objection.

Reply:
Similar problem is in `db_hash_process_pending` as this is a general issue. I will remove nack as I guess maintainers of this code will have chance to see if this is a real problem in there or not.